File: api/src/main/java/net/blueshell/api/job/calendar/AddEventToCalendarJob.java
package net.blueshell.api.job.calendar;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.blueshell.api.model.event.Event;
import net.blueshell.api.service.CalendarService;
import net.blueshell.api.service.event.EventService;
import org.springframework.retry.annotation.Backoff;
import org.springframework.retry.annotation.Recover;
import org.springframework.retry.annotation.Retryable;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReentrantLock;

@Service
@Slf4j
@RequiredArgsConstructor
public class AddEventToCalendarJob {

    private static final ConcurrentHashMap<Long, ReentrantLock> locks = new ConcurrentHashMap<>();
    private static final ConcurrentHashMap<String, Boolean> processing = new ConcurrentHashMap<>();

    private final CalendarService calendarService;
    private final EventService eventService;

    @Async
    @Retryable(retryFor = {Exception.class, UncheckedIOException.class}, maxAttempts = 3,
            backoff = @Backoff(delay = 2000, multiplier = 2))
    public CompletableFuture<Void> add(Long eventId) {
        String key = key("add", eventId);
        if (processing.putIfAbsent(key, true) != null) return CompletableFuture.completedFuture(null);

        ReentrantLock lock = locks.computeIfAbsent(eventId, k -> new ReentrantLock());
        try {
            lock.lock();
            Event e = eventService.findById(eventId);
            if (e == null) {
                log.warn("Add skipped: eventId {} not found", eventId);
                return CompletableFuture.completedFuture(null);
            }
            if (!e.isApproved()) {
                log.info("Add skipped: eventId {} not approved", eventId);
                return CompletableFuture.completedFuture(null);
            }

            calendarService.add(e);
            eventService.update(e);
            log.info("Added eventId {} to Google Calendar as {}", eventId, e.getGoogleId());
            return CompletableFuture.completedFuture(null);
        } catch (IOException ex) {
            throw new UncheckedIOException(ex);
        } finally {
            lock.unlock();
            processing.remove(key);
        }
    }

    @Recover
    public CompletableFuture<Void> recover(Exception ex, Long eventId) {
        processing.remove(key("add", eventId));
        log.error("Giving up add for eventId {}: {}", eventId, ex.getMessage(), ex);
        return CompletableFuture.completedFuture(null);
    }

    private String key(String op, Long id) {
        return op + "_" + id + "_" + (System.currentTimeMillis() / 10000);
    }
}

File: api/src/main/kotlin/net/blueshell/api/job/calendar/AddEventToCalendarJob.kt
package net.blueshell.api.job.calendar

import net.blueshell.api.service.CalendarService
import net.blueshell.api.service.event.EventService
import org.slf4j.LoggerFactory
import org.springframework.retry.annotation.Backoff
import org.springframework.retry.annotation.Recover
import org.springframework.retry.annotation.Retryable
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import java.io.IOException
import java.io.UncheckedIOException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.locks.ReentrantLock

@Service
class AddEventToCalendarJob(
    private val calendarService: CalendarService,
    private val eventService: EventService
) {
    @Async
    @Retryable(
        retryFor = [Exception::class, UncheckedIOException::class],
        maxAttempts = 3,
        backoff = Backoff(delay = 2000, multiplier = 2)
    )
    fun add(eventId: Long?): CompletableFuture<Void?> {
        val key = key("add", eventId)
        if (processing.putIfAbsent(key, true) != null) return CompletableFuture.completedFuture<Void?>(null)

        val lock: ReentrantLock = locks.computeIfAbsent(eventId) { k: Long? -> ReentrantLock() }
        try {
            lock.lock()
            val e = eventService.findById(eventId)
            if (e == null) {
                log.warn("Add skipped: eventId {} not found", eventId)
                return CompletableFuture.completedFuture<Void?>(null)
            }
            if (!e.approved) {
                log.info("Add skipped: eventId {} not approved", eventId)
                return CompletableFuture.completedFuture<Void?>(null)
            }

            calendarService.add(e)
            eventService.update(e)
            log.info("Added eventId {} to Google Calendar as {}", eventId, e.getGoogleId())
            return CompletableFuture.completedFuture<Void?>(null)
        } catch (ex: IOException) {
            throw UncheckedIOException(ex)
        } finally {
            lock.unlock()
            processing.remove(key)
        }
    }

    @Recover
    fun recover(ex: Exception, eventId: Long?): CompletableFuture<Void?> {
        processing.remove(key("add", eventId))
        log.error("Giving up add for eventId {}: {}", eventId, ex.message, ex)
        return CompletableFuture.completedFuture<Void?>(null)
    }

    private fun key(op: String?, id: Long?): String {
        return op + "_" + id + "_" + (System.currentTimeMillis() / 10000)
    }

    companion object {
        private val log = LoggerFactory.getLogger(AddEventToCalendarJob::class.java)
        private val locks = ConcurrentHashMap<Long?, ReentrantLock>()
        private val processing = ConcurrentHashMap<String?, Boolean?>()
    }
}
